network/p2p/cache/node_blocklist_wrapper.go

```diff
@@ -0,0 +1,191 @@
+// filter for removing ejected nodes).
+func (w *NodeBlocklistWrapper) Identities(filter flow.IdentityFilter) flow.IdentityList {
+	identities := w.identityProvider.Identities(filter)
+	if len(identities) == 0 {
+		return identities
+	}
+
+	// Iterate over all returned identities and set ejected flag to true. We
+	// copy both the return slice and identities of blocked nodes to avoid
+	// any possibility of accidentally modifying the wrapped IdentityProvider
+	idtx := make(flow.IdentityList, 0, len(identities))
+	w.m.RLock()
+	for _, identity := range identities {
+		if w.blocklist.Includes(identity.NodeID) {
+			var i flow.Identity = *identity // shallow copy is sufficient, because `Ejected` flag is in top-level struct
+			i.Ejected = true
+			idtx = append(idtx, &i)
+		} else {
+			idtx = append(idtx, identity)
+		}
+	}
+	w.m.RUnlock()
+	return idtx
+}
+
+// ByNodeID returns the full identity for the node with the given Identifier,
+// where Identifier is the way the protocol refers to the node. The function
+// has the same semantics as a map lookup, where the boolean return value is
+// true if and only if Identity has been found, i.e. `Identity` is not nil.
+// Caution: function returns include ejected nodes. Please check the `Ejected`
+// flag in the identity.
+func (w *NodeBlocklistWrapper) ByNodeID(identifier flow.Identifier) (*flow.Identity, bool) {
+	identity, b := w.identityProvider.ByNodeID(identifier)
+	return w.applyBlocklist(identity), b
+}
+
+// applyBlocklist checks whether the node with the given identity is on the `blocklist`.
+// Shortcuts:
+//   - If the node's identity is nil, there is nothing to do because we don't generate identities here.
+//   - If the node is already ejected, we don't have to check the blocklist.
+func (w *NodeBlocklistWrapper) applyBlocklist(identity *flow.Identity) *flow.Identity {
+	if identity == nil || identity.Ejected {
+		return identity
+	}
+
+	w.m.RLock()
+	isBlocked := w.blocklist.Includes(identity.NodeID)
+	w.m.RUnlock()
+	if !isBlocked {
+		return identity
+	}
+
+	// For blocked nodes, we want to return their `Identity` with the `Ejected` flag
+	// set to true. Caution: we need to copy the `Identity` before we override `Ejected`, as we
+	// would otherwise potentially change the wrapped IdentityProvider.
+	var i flow.Identity = *identity // shallow copy is sufficient, because `Ejected` flag is in top-level struct
+	i.Ejected = true
+	return &i
+}
+
+// ByPeerID returns the full identity for the node with the given peer ID,
+// where ID is the way the libP2P refers to the node. The function
+// has the same semantics as a map lookup, where the boolean return value is
+// true if and only if Identity has been found, i.e. `Identity` is not nil.
+// Caution: function returns include ejected nodes. Please check the `Ejected`
+// flag in the identity.
+func (w *NodeBlocklistWrapper) ByPeerID(p peer.ID) (*flow.Identity, bool) {
+	identity, b := w.identityProvider.ByPeerID(p)
```

Comment:
Both methods [`ByNodeID`](https://github.com/onflow/flow-go/blob/f1bdd42a29a882a0674d36cbd286411c6f0f01c0/network/p2p/cache/node_blocklist_wrapper.go#L131-L134) as well as [`ByPeerID`](https://github.com/onflow/flow-go/blob/f1bdd42a29a882a0674d36cbd286411c6f0f01c0/network/p2p/cache/node_blocklist_wrapper.go#L166-L169) must implement this check. 
In my implementation, both methods call into `applyBlacklist`. If I was to move this check out, I would need to unnecessarily duplicate these checks in [`ByNodeID`](https://github.com/onflow/flow-go/blob/f1bdd42a29a882a0674d36cbd286411c6f0f01c0/network/p2p/cache/node_blocklist_wrapper.go#L131-L134) and [`ByPeerID`](https://github.com/onflow/flow-go/blob/f1bdd42a29a882a0674d36cbd286411c6f0f01c0/network/p2p/cache/node_blocklist_wrapper.go#L166-L169). My preference is to avoid unnecessary code duplication. 

Another way to read your comment could maybe be that the method name `applyBlacklist` is unclear. I have renamed the method to `setEjectedIfBlocked`.

Reply:
Please also see [my comment below](https://github.com/onflow/flow-go/pull/3344#discussion_r990570878), why I prefer to only look at the returned identity, irrespective of the boolean value (defensive programming)